src/services/Sync/index.ts

```diff
@@ -0,0 +1,270 @@
+import * as R from 'ramda';
+
+import { CoreModuleName } from '../../utils/createCoreModule';
+import store from '@/store';
+import { CoreModuleState, CoreRow, StoreRow } from '../../store/types';
+import CoreAPI from '../../utils/CoreAPI/index';
+import { getModulePath, CoreModuleNames } from '@/utils/createCoreModule/utils';
+
+/**
+ * Promise representing the job taking place
+ */
+type Job = Promise<boolean>;
+
+type Jobs = Promise<boolean[]>;
+
+interface CurrentModuleJobs {
+  [key: string]: Job | undefined;
+}
+
+/** Instance of the sync service */
+export interface SyncServiceInstance {
+
+  /**
+   * Whether the service is currently syncing or not
+   */
+  readonly syncing: boolean;
+
+  /**
+   * Start the sync service
+   */
+  start(): void;
+
+  /**
+   * Sync all the core modules
+   *
+   * @returns A promise that resolves when all modules have successfully synced
+   */
+  syncAll(): Jobs;
+
+  /**
+   * Sync the specified module with the core
+   *
+   * @returns A promise that resolves when module has successfully synced
+   */
+  syncModule(module: CoreModuleName): Job;
+}
+
+/** Sync service module */
+export interface SyncService {
+  /** Get instance of sync service */
+  (): SyncServiceInstance;
+
+  /** Stop the sync service from running */
+  stop(): Promise<void>;
+}
+
+/**
+ * Keeps track of the current sync interval id
+ */
+let intervalId = -1;
+
+/*
+* A symbol is a primitive type that is guaranteed to be unique. That means
+* the only thing that will match this symbol is this exact symbol right here.
+* The only way to access the instance of SyncServiceInstance is using this symbol
+* defined here.
+*
+* This is useful to mock out the SyncService during tests
+*/
+const INSTANCE_ACCESSOR = Symbol('Accessor for instance of SyncService');
+
+/** Default number of seconds between automatic sync */
+const DEFAULT_FREQUENCY = 30;
+
+/** Number of milliseconds in a second */
+const TO_MILLISECONDS = 1000;
+
+/** The sync frequency for automatic sync */
+const SYNC_FREQUENCY = TO_MILLISECONDS * (Number(process.env.SYNC_FREQUENCY) || DEFAULT_FREQUENCY);
+
+/**
+ * Keep track of the current module jobs
+ *
+ * Represents active modules that are currently syncing
+ */
+const activeModuleJobs = {} as CurrentModuleJobs;
+
+/**
+ * Type guard to determine if a module is an active job
+ *
+ * @param a Object to test
+ */
+function isJob<T>(a: Promise<T> | undefined): a is Promise<T> {
+  return (typeof a !== 'undefined');
+}
+
+/**
+ * Determine whether the sync service interval is active or not
+ *
+ * @param intervalId ID for sync service interval
+ */
+function isRunning(intervalId: number) {
+  return !!~intervalId;
```

Comment:
what's ~ mean?

Reply:
It's the bitwise NOT operator, which honestly I don't even remember bitwise stuff. I just know a neat trick is that the bitwise NOT of -1 is 0, while the bitwise NOT of anything else is not 0. It's just a common shorthand for checking if something is -1 (often for indexOf or stuff like that)